src/vsr/replica.zig

```diff
@@ -1143,68 +1235,328 @@ pub fn Replica(
+                self.journal.dirty.count,
+                self.journal.faulty.count,
+            });
+
+            self.state_machine.prepare_timestamp = self.journal.header_with_op(self.op).?.timestamp;
+            // `state_machine.commit_timestamp` is updated as messages are committed.
+            self.commit_ops(commit);
+            self.repair();
+
+            self.reset_quorum_recovery_response();
         }
 
+        /// If the requested prepare has been guaranteed by this replica:
+        /// * Read the prepare from storage, and forward it to the replica that requested it.
+        /// * Otherwise send no reply — it isn't safe to nack.
+        /// If the requested prepare has *not* been guaranteed by this replica, then send a nack.
+        ///
+        /// A prepare is considered "guaranteed" by a replica if that replica has acknowledged it
+        /// to the cluster. The cluster sees the replica as an underwriter of a guaranteed
+        /// prepare. If a guaranteed prepare is found to by faulty, the replica must repair it
+        /// to restore durability.
         fn on_request_prepare(self: *Self, message: *const Message) void {
             if (self.ignore_repair_message(message)) return;
 
+            assert(self.replica_count > 1);
             assert(self.status == .normal or self.status == .view_change);
             assert(message.header.view == self.view);
             assert(message.header.replica != self.replica);
 
             const op = message.header.op;
-            var checksum: ?u128 = message.header.context;
-            if (self.leader_index(self.view) == self.replica and checksum.? == 0) checksum = null;
-
-            if (self.journal.entry_for_op_exact_with_checksum(op, checksum)) |entry| {
-                assert(entry.op == op);
-                assert(checksum == null or entry.checksum == checksum.?);
+            const slot = self.journal.slot_for_op(op);
+            const checksum: ?u128 = switch (message.header.timestamp) {
+                0 => null,
+                1 => message.header.context,
+                else => unreachable,
+            };
 
-                if (!self.journal.dirty.bit(op)) {
-                    assert(!self.journal.faulty.bit(op));
+            if (message.header.timestamp == 0) {
+                assert(self.leader_index(self.view) == self.replica);
+            }
 
-                    log.debug("{}: on_request_prepare: op={} checksum={} reading", .{
+            if (self.leader_index(self.view) != self.replica) {
+                // Only the leader may respond to `request_prepare` messages without a checksum.
+                if (checksum == null) return;
+            }
+
+            // Try to serve the message directly from the pipeline.
+            // This saves us from going to disk. And we don't need to worry that the WAL's copy
+            // of an uncommitted prepare is lost/corrupted.
+            var it = self.pipeline.iterator();
```

Comment:
We can use the new `pipeline_prepare_for_op_and_checksum()` helper that I added here as well, if we tweak the `checksum` arg to be optional for the helper.

Reply:
That also gives the benefit of not searching the pipeline at all if the op is outside either side.

Reply:
I realize now that we can always index directly into the pipeline by op and then check the checksum, provided we assert that the pipeline latches onto our current `commit_min` number.